Created the MS-Calculator microservice.

futind opened the pull request:
Defined the DTOs, defined and implemented endpoints for "/offers" and "/calc".

futind committed:
Created the MS-Calculator microservice.
Defined the DTOs, defined and implemented endpoints for "/offers" and "/calc".

KonstantinSmelov commented on MS-Calculator/src/main/java/ru/neoflex/mscalculator/controller/CalculatorController.java:
```diff
@@ -0,0 +1,41 @@
+
+import java.util.List;
+
+@RestController
+@RequestMapping
+public class CalculatorController {
+
```
В задании сказано, что пути до эндпоинтов /calculator/offers и /calculator/calc. У тебя просто offers и calc.

KonstantinSmelov commented on MS-Calculator/src/main/java/ru/neoflex/mscalculator/controller/CalculatorController.java:
```diff
@@ -0,0 +1,41 @@
+    }
+
+    @PostMapping(path = "/offers")
+    public List<LoanOfferDto> getOffers(@Valid @RequestBody @NotNull LoanStatementRequestDto loanStatementRequestDto) {
```
Когда я у себя запускал твой код, то аннотации Valid и NotNull, используемые одновременно, не работают как надо. Вроде бы надо брать аннотацию Validated, если ты хочешь использовать одновременно с NotNull

KonstantinSmelov commented on MS-Calculator/src/main/java/ru/neoflex/mscalculator/controller/CalculatorController.java:
```diff
@@ -0,0 +1,41 @@
+        return calculatorService.getCredit(scoringDataDto);
+    }
+
+    @ExceptionHandler(IllegalArgumentException.class)
+    @ResponseStatus(HttpStatus.BAD_REQUEST)
+    public String handleIAException(IllegalArgumentException e) {
+        return "Illegal argument: " + e.getMessage();
+    }
```
Надо создать пакет, к примеру, exception и уже там создать класс, который будет хэндлить исключения (подсказка - @ControllerAdvice). Нечего этому добру в классе контроллера делать.

KonstantinSmelov commented on MS-Calculator/src/main/java/ru/neoflex/mscalculator/service/CalculatorService.java:
```diff
@@ -0,0 +1,258 @@
+    }
+
+    // prescoring
+    public List<LoanOfferDto> getOffers(LoanStatementRequestDto loanStatementRequestDto) {
+
+        LoanOfferDto withoutInsuranceNotClient = createLoanOfferDto(loanStatementRequestDto.getAmount(),
+                loanStatementRequestDto.getTerm(),
+                false, false);
+        LoanOfferDto withoutInsuranceAClient = createLoanOfferDto(loanStatementRequestDto.getAmount(),
+                loanStatementRequestDto.getTerm(),
+                false, true);
+        LoanOfferDto withInsuranceNotClient = createLoanOfferDto(loanStatementRequestDto.getAmount(),
+                loanStatementRequestDto.getTerm(),
+                true, false);
+        LoanOfferDto withInsuranceAClient = createLoanOfferDto(loanStatementRequestDto.getAmount(),
+                loanStatementRequestDto.getTerm(),
+                true, true);
+        List<LoanOfferDto> loanOfferDtoList = Arrays.asList(withoutInsuranceNotClient, withoutInsuranceAClient,
+                withInsuranceNotClient, withInsuranceAClient);
+        loanOfferDtoList.sort(new RateComparator());
+        return loanOfferDtoList;
+    }
```
На мой вкус и по логике избыточно много созданий новых переменных. Зачем создавать 4-ре переменные LoanOfferDto?!
    Понятно, что Garbage Collector их почистит, но это всё же лишняя работа.
    Создаёшь List<LoanOfferDto> и в него добавляешь (add()) четыре createLoanOfferDto(), которые в свою очередь возвращают нужные LoanOfferDto.  

    И ещё. Больше придирка, но всё же я уверен, что будет лучше читаться, если в createLoanOfferDto() не передавать loanStatementRequestDto.getAmount(), и loanStatementRequestDto.getTerm(), а просто передать саму ДТО. А уже внутри функции доставать из неё то, что тебе надо.

KonstantinSmelov commented on MS-Calculator/src/main/java/ru/neoflex/mscalculator/service/CalculatorService.java:
```diff
@@ -0,0 +1,258 @@
+
+        BigDecimal psk = calculatePsk(monthlyPayment, scoringDataDto.getTerm());
+
+        return new CreditDto(amountWithInsurance.setScale(2, RoundingMode.HALF_EVEN),
+                             scoringDataDto.getTerm(),
+                             monthlyPayment.setScale(2, RoundingMode.HALF_EVEN),
+                             adjustedRate,
+                             psk.setScale(2, RoundingMode.HALF_EVEN),
+                             scoringDataDto.getIsInsuranceEnabled(),
+                             scoringDataDto.getIsSalaryClient(),
+                             paymentSchedule);
```
Я бы создавал объект, используя паттерн "билдер" Тем более Ломбок его сам реализует для твоего LoanOfferDto, если ты добавишь аннотацию Builder.
Конструктор обязывает тебе инициализировать все поля (ну или явно ставить null, для тех, которые ты не хочешь инициализировать на текущем этапе). А при создании объекта через билдер, ты задаёшь только те поля которые тебя нужны.
В данном случае это неактуально, т.к. ты инициализируешь все поля, но частенько это не нужно делать сразу в момент создания объекта. Используя в курсовом проекте билдер, ты как минимум набьёшь себе руку.

KonstantinSmelov commented on MS-Calculator/src/main/java/ru/neoflex/mscalculator/service/CalculatorService.java:
```diff
@@ -0,0 +1,258 @@
+        BigDecimal maxAmount = scoringDataDto.getEmployment().getSalary().multiply(BigDecimal.valueOf(24));
+        int age = (int) ChronoUnit.YEARS.between(scoringDataDto.getBirthdate(), LocalDate.now());
+
+        if (scoringDataDto.getEmployment().getEmploymentStatus() == EmploymentStatus.NOT_EMPLOYED ||
+            scoringDataDto.getEmployment().getWorkExperienceTotal() < 18 ||
+            scoringDataDto.getEmployment().getWorkExperienceCurrent() < 3 ||
+            age < 20 || age > 65 ||
+            scoringDataDto.getAmount().add(insurancePayment).compareTo(maxAmount) > 0) {
+            throw new IllegalArgumentException("Credit application is denied");
```
Видится, что отказ надо отрабатывать максимально рано. А то у тебя код уже поработал, что-то повычислял, посравнивал (утилизировал процессорное время и ресурсы сервера), а потом - бац - а это оказалось всё впустую, т.к. отказ!
Плюс к этому, мне не нравится эта ошибка. Аргументы абсолютно легальные. Другое дело, что они банку не подходят. Создай свою ошибку и обработай в @ControllerAdvice.

Задание со звёздочкой (не обязательно): реализовать свой Эксепшен не просто в виде "какие-то ваши данные не удовлетворяют условиям - до свидания", а массивом ошибок, чтобы было понятно какие конкретно параметры банк не устраивают.

KonstantinSmelov commented on MS-Calculator/src/main/java/ru/neoflex/mscalculator/service/CalculatorService.java:
```diff
@@ -0,0 +1,258 @@
+
+        BigDecimal insuredAmount = requestedAmount.add(insurancePayment);
+
+        // Как лучше округлять?
```
Т.к. проект учебный - я думаю, неважно, если в разумных пределах. В реальном проекте тебе аналитик скажет как округлять (а если не скажет, ты у него спросишь)

KonstantinSmelov commented on MS-Calculator/src/main/java/ru/neoflex/mscalculator/service/CalculatorService.java:
```diff
@@ -0,0 +1,258 @@
+                                rate, isInsuranceClient, isSalaryClient);
+    }
+
+    private final BigDecimal baseRate;
+    private final BigDecimal insuranceRate;
```
Что тут происходит?? Объявление переменных перед функциями, а не после :)

KonstantinSmelov commented on MS-Calculator/src/main/java/ru/neoflex/mscalculator/service/CalculatorService.java:
```diff
@@ -0,0 +1,258 @@
+
+}
+
+class RateComparator implements Comparator<LoanOfferDto> {
+    @Override
+    public int compare(LoanOfferDto o1, LoanOfferDto o2) {
+        return o1.getRate().compareTo(o2.getRate());
+    }
+}
```
Не уверен, что хорошая идея создавать несколько классов в одном файле. Создай пакет, к примеру, util и там уже плоди подобные технические классы.

futind committed:
Fixes, Logging, Documentation and Unit-tests.
Fixed everything that was pointed out in the previous PR. Added logging, documentation using Swagger and Unit-tests using JUnit.

KonstantinSmelov commented on MS-Calculator/src/main/java/ru/neoflex/mscalculator/service/CalculatorService.java:
```diff
@@ -1,88 +1,141 @@
+        loanOfferDtoList.add(createLoanOfferDto(loanStatementRequestDto,
+                true, true));
+
+        loanOfferDtoList.sort(new RateComparator().reversed());
```
зачем .reversed()?  по заданию надо представить предложения от лучшему к худшему (лучше - меньше ставка).

KonstantinSmelov commented on MS-Calculator/src/main/java/ru/neoflex/mscalculator/dtos/CreditDto.java:
```diff
@@ -1,41 +1,72 @@
 @Getter
 @Setter
+@NoArgsConstructor
+@AllArgsConstructor
```
Тут, по моему, во-первых, не все конструкторы в коде используются. А во-вторых, прекрасно работает с одной аннотацией @ Data, которая включает в себя с аннотацию сеттер, геттер, и ещё несколько.

KonstantinSmelov commented on MS-Calculator/src/main/java/ru/neoflex/mscalculator/dtos/CreditDto.java:
```diff
@@ -1,41 +1,72 @@
     }
-
+    @NotNull
+    @Schema(description = "Сумма займа в рублях")
```
В аннотации Схема есть параметр example. Если его заполнить, то в Сваггере у тебя подставится то значение в модель, которое будет указано. Очень удобно. 
Сейчас у тебя подставляется биллиберда типа 
"firstName": "IgSsSkKnMElkVgJjzXMDsYMtJumQJy"
"email": "x9y?TD4}E^/Vr4%r6AW*8FAF3{rOFDwxtqCGgfWBlK/~.NtGl$aWe^j#Vr2WMu+JemUJDD@i3kXAcr5ywoZJUqISSYwRwdQIYLYV9-KhEc-hobhc7AsDhkx7bRwMczGDRm9",

futind committed:
Annotation changes and Swagger examples
